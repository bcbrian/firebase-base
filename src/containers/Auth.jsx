/*
  This is the auth container
  it uses a render prop
  it calls the render prop, children prop, with user
  if user is null then they are not authenticated
  it also calls it with the authentication functions
  signUp, signIn, signOut
*/

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { auth } from '../firebase';
import { handleSignUp, handleSignIn, handleSignOut } from '../firebase/auth';

export default class Auth extends Component {
  constructor(props) {
    super(props);

    this.state = {
      user: null
    };

    this.unsubscribeFromAuthStateChange = () => {};
  }

  componentDidMount() {
    this.unsubscribeFromAuthStateChange = auth.onAuthStateChanged(user => {
      this.setState(() => ({
        user
      }));
    });
  }

  componentWillUnmount() {
    this.unsubscribeFromAuthStateChange();
  }

  render() {
    const { user } = this.state;
    if (this.props.children) {
      return this.props.children({
        user,
        handleSignUp,
        handleSignIn,
        handleSignOut
      });
    } else if (this.props.render) {
      return this.props.render({
        user,
        handleSignUp,
        handleSignIn,
        handleSignOut
      });
    }
    return <div>You did not provide a render or children prop.</div>;
  }
}

Auth.propTypes = {
  children: PropTypes.func,
  render: PropTypes.func
};

Auth.defaultProps = {
  children: null,
  render: null
};
