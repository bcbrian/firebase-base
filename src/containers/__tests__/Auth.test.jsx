import React from 'react';
import renderer from 'react-test-renderer';
import Auth from '../Auth';
import { auth } from '../../firebase';
import * as userAuthFunctions from '../../firebase/auth';

jest.mock('firebase');

describe('Auth Component', () => {
  let user;
  let handleSignUp;
  let handleSignIn;
  let handleSignOut;
  beforeEach(() => {
    user = {
      email: 'test@test.com'
    };
    auth.onAuthStateChanged = jest.fn(callback => callback(user));

    handleSignUp = jest.fn();
    handleSignIn = jest.fn();
    handleSignOut = jest.fn();

    userAuthFunctions.handleSignUp = handleSignUp;
    userAuthFunctions.handleSignIn = handleSignIn;
    userAuthFunctions.handleSignOut = handleSignOut;
  });
  it('returns the user and functions to children', () => {
    const childRenderProp = jest.fn(() => <div>Children Render Prop</div>);
    const AuthComp = renderer.create(<Auth>{childRenderProp}</Auth>);
    expect(childRenderProp).toHaveBeenCalledTimes(2);
    expect(childRenderProp).lastCalledWith({
      user,
      handleSignUp: userAuthFunctions.handleSignUp,
      handleSignIn: userAuthFunctions.handleSignIn,
      handleSignOut: userAuthFunctions.handleSignOut
    });
    expect(AuthComp.toJSON()).toMatchSnapshot();
  });

  it('returns the user and functions to render', () => {
    const renderProp = jest.fn(() => <div>Render Prop</div>);
    const AuthComp = renderer.create(<Auth render={renderProp} />);
    expect(renderProp).toHaveBeenCalledTimes(2);
    expect(renderProp).lastCalledWith({
      user,
      handleSignUp: userAuthFunctions.handleSignUp,
      handleSignIn: userAuthFunctions.handleSignIn,
      handleSignOut: userAuthFunctions.handleSignOut
    });
    expect(AuthComp.toJSON()).toMatchSnapshot();
  });

  it('returns develop warning if there are not children or reander props', () => {
    const AuthComp = renderer.create(<Auth />);
    expect(AuthComp.toJSON()).toMatchSnapshot();
  });
});
