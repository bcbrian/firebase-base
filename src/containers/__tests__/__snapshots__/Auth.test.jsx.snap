// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Component returns develop warning if there are not children or reander props 1`] = `
<div>
  You did not provide a render or children prop.
</div>
`;

exports[`Auth Component returns the user and functions to children 1`] = `
<div>
  Children Render Prop
</div>
`;

exports[`Auth Component returns the user and functions to render 1`] = `
<div>
  Render Prop
</div>
`;
